src/librustc_mir/interpret/place.rs

```diff
@@ -869,10 +868,10 @@ where
         // We do NOT compare the types for equality, because well-typed code can
         // actually "transmute" `&mut T` to `&T` in an assignment without a cast.
         assert!(
-            src.layout.layout == dest.layout.layout,
-            "Layout mismatch when copying!\nsrc: {:#?}\ndest: {:#?}",
-            src,
-            dest
+            mir_assign_valid_types(src.layout.ty, dest.layout.ty),
+            "type mismatch when copying!\nsrc: {:?},\ndest: {:?}",
+            src.layout.ty,
```

Comment:
why not print the entire `Layout` here? It may make it easier to debug the situation.

Reply:
In every single case of this assertion failing that I encountered so far, the layout was actually identical except for the type. Printing the entire thing drowns the useful information.

But I can add the layout as well, if you wish.

Reply:
Nah, we can always change it if we start noticing anything